File: oneground/jobs.py
"""What a job is: one CLI invocation, its state, and what it left behind.

Task 046 step 3. `docs/INTERFACE.md` §4.2 makes each stage a job, and §2 makes
every job exactly a CLI invocation that a test can replay.

WHY THIS IS NOT IN `oneground/lab/`
-----------------------------------
The lab's guard permits exactly one served module to open a file for writing,
and that module writes requirements files. A job list that the server could
write would put a second write path inside the served package, so the split
is: **the server reads jobs, the supervisor writes them** (§4, a second
process). This module holds the record, the state machine and the reading;
the writing belongs to the process that runs things.

That is the same ruling as *the UI does not launch pod sessions*, one level
down: a server that may not import the measuring packages cannot run them,
and a server that may not write cannot own the record of what ran.

THE STATES, AND WHY THE TRANSITIONS ARE CHECKED
-----------------------------------------------
Six states, and the set is closed. `refused` is not a kind of `failed`: a
refusal is the tool declining to run something, with a reason and a remedy,
and a failure is something breaking. `docs/INTERFACE.md` §4.4 requires the
refusal be shown verbatim, and collapsing the two would make the most useful
message the tool produces into an error string.

The transitions are checked because the states are what the page draws from.
A job that went from `done` back to `running` would render a finished run as
in progress, and nothing downstream would notice -- the same shape as a
receipt that says something it should not.
"""

import json
import os
import re

#: Queued and running are live; the rest are terminal.
QUEUED = "queued"
RUNNING = "running"
DONE = "done"
FAILED = "failed"
REFUSED = "refused"
CANCELLED = "cancelled"

STATES = (QUEUED, RUNNING, DONE, FAILED, REFUSED, CANCELLED)
TERMINAL = (DONE, FAILED, REFUSED, CANCELLED)

#: What may follow what. A refusal can happen before the work starts -- an
#: unpinned environment is refused at parse time -- so `queued -> refused` is
#: legal and is not a failure.
TRANSITIONS = {
    QUEUED: (RUNNING, REFUSED, CANCELLED),
    RUNNING: (DONE, FAILED, REFUSED, CANCELLED),
    DONE: (), FAILED: (), REFUSED: (), CANCELLED: (),
}

#: The stages a job may be. Every pod operation except the create: `up` is a
#: terminal, because a scriptable page is `--yes` with a nicer surface
#: (`docs/INTERFACE.md` §4.3).
STAGES = ("characterize", "simulate", "verify", "report", "chunk", "propose",
          "pod plan", "pod status", "pod watch", "pod fetch")

#: Never a job, and named rather than merely absent, because the reason is a
#: ruling rather than an omission.
NOT_A_JOB = {
    "pod up": "the one operation that spends money keeps the one boundary "
              "that cannot be automated by accident; the UI shows the card "
              "and hands over the command",
}

_ID = re.compile(r"^[0-9]{8}T[0-9]{6}Z-[a-z0-9-]{1,40}-[0-9a-f]{8}$")


class JobError(ValueError):
    """A job record is malformed, or a transition is not one the states
    allow."""


class Job:
    """One invocation, its state, and what it left behind.

    `invocation` is the field everything else hangs off: it is what
    `provenance.record_invocation` wrote into the receipts this job produced,
    and what `oneground.replay` re-runs to check them.
    """

    __slots__ = ("id", "stage", "invocation", "state", "workdir", "log",
                 "started", "ended", "refusal", "exit_code", "partial")

    def __init__(self, id, stage, invocation, workdir, state=QUEUED,
                 log=None, started=None, ended=None, refusal=None,
                 exit_code=None, partial=False):
        if stage not in STAGES:
            raise JobError(
                f"{stage!r} is not a stage a job may be. Stages: "
                + ", ".join(STAGES)
                + "".join(f"; {k} is never a job because {v}"
                          for k, v in NOT_A_JOB.items()))
        if state not in STATES:
            raise JobError(f"{state!r} is not a state; states: "
                           + ", ".join(STATES))
        if not invocation:
            raise JobError(
                "a job is a CLI invocation and this one names none, so "
                "nothing it produced could be replayed")
        self.id = id
        self.stage = stage
        self.invocation = list(invocation)
        self.state = state
        self.workdir = workdir
        self.log = log
        self.started = started
        self.ended = ended
        #: The CLI's own words, verbatim. Never rephrased and never
        #: summarised -- `docs/INTERFACE.md` §4.4.
        self.refusal = refusal
        self.exit_code = exit_code
        #: A cancelled job's outputs are kept and marked, never presented as
        #: complete. The mark travels with the record rather than being
        #: inferred from the state, because a reader of one artifact does not
        #: have the state.
        self.partial = bool(partial)

    # -- the state machine --------------------------------------------------
    def may_become(self, state):
        return state in TRANSITIONS.get(self.state, ())

    def become(self, state, *, refusal=None, exit_code=None, ended=None,
               partial=None):
        """Move to `state`, or refuse to.

        Refusing an illegal transition is the point. The states are what the
        page draws from, so a job that went from `done` back to `running`
        would render a finished run as in progress and nothing downstream
        would notice.
        """
        if not self.may_become(state):
            allowed = TRANSITIONS.get(self.state, ())
            raise JobError(
                f"a {self.state} job cannot become {state}; from "
                f"{self.state} a job may become "
                + (", ".join(allowed) if allowed else
                   "nothing: it is a terminal state"))
        if state == REFUSED and not refusal:
            raise JobError(
                "a refused job carries the CLI's refusal verbatim; refusing "
                "without one would lose the most useful thing the tool "
                "produces")
        self.state = state
        if refusal is not None:
            self.refusal = refusal
        if exit_code is not None:
            self.exit_code = exit_code
        if ended is not None:
            self.ended = ended
        if partial is not None:
            self.partial = bool(partial)
        return self

    @property
    def live(self):
        return self.state not in TERMINAL

    # -- the record ---------------------------------------------------------
    def to_dict(self):
        return {"id": self.id, "stage": self.stage,
                "invocation": list(self.invocation), "state": self.state,
                "workdir": self.workdir, "log": self.log,
                "started": self.started, "ended": self.ended,
                "refusal": self.refusal, "exit_code": self.exit_code,
                "partial": self.partial}

    @classmethod
    def from_dict(cls, d):
        if not isinstance(d, dict):
            raise JobError(f"a job record is a mapping, not a "
                           f"{type(d).__name__}")
        missing = [k for k in ("id", "stage", "invocation", "state")
                   if d.get(k) in (None, "")]
        if missing:
            raise JobError("a job record is missing " + ", ".join(missing))
        return cls(id=d["id"], stage=d["stage"], invocation=d["invocation"],
                   workdir=d.get("workdir"), state=d["state"],
                   log=d.get("log"), started=d.get("started"),
                   ended=d.get("ended"), refusal=d.get("refusal"),
                   exit_code=d.get("exit_code"),
                   partial=d.get("partial", False))

    def __repr__(self):                                # pragma: no cover
        return f"<Job {self.id} {self.stage} {self.state}>"


def valid_id(value):
    """Whether a job id has the shape this project writes.

    Checked because an id reaches the filesystem as a log file's name. It is
    a timestamp, the stage, and eight hex -- sortable, readable in a
    directory listing, and containing nothing a path would have to escape.
    """
    return bool(isinstance(value, str) and _ID.match(value))


#: The job list's file name, inside the runs directory the session serves.
JOBS_NAME = "jobs.json"


def read(runs_dir):
    """Every job the supervisor has recorded, oldest first.

    Reading only. The server calls this; nothing here writes, which is what
    keeps the served package's single write path single.

    An absent file is an empty list rather than an error: a runs directory
    that has never had a job is the ordinary case, and refusing it would
    make the page's first visit a failure.
    """
    path = os.path.join(runs_dir, JOBS_NAME)
    if not os.path.isfile(path):
        return []
    try:
        with open(path, encoding="utf-8") as f:
            raw = json.load(f)
    except (OSError, ValueError) as e:
        raise JobError(f"{JOBS_NAME} could not be read: {e}") from None
    if not isinstance(raw, list):
        raise JobError(f"{JOBS_NAME} holds a {type(raw).__name__}, not a "
                       "list of jobs")
    return [Job.from_dict(d) for d in raw]

File: oneground/lab/static/jobs.js
// The jobs page: what has run, what is running, and what it left behind.
//
// Task 046 steps 3 and 4. Its own file for the same reason `compose.js` is:
// `ui.js` opens by saying there is no POST in it, and the read half and the
// write half are kept apart in the code and in the page.
//
// WHY THIS PAGE TALKS TO TWO SERVERS
// ----------------------------------
// It reads from the lab server and it *asks* the supervisor. That is not an
// accident of layout, it is the shape of a ruling: the lab server's own
// header has promised since it was written that it makes no request of its
// own, and it has no write path. So it cannot enqueue a job and it will not
// forward one. The page is told where the supervisor is and does the asking
// itself, with the supervisor's token.
//
// The supervisor permits exactly one origin -- this page's -- and the token
// is still required on every request. **CORS says which page the browser
// will let read a reply; it says nothing about who may ask.** If the next
// person here starts treating the allowance as authentication, that is the
// sentence they should meet first.
//
// WHAT IS DRAWN AND WHAT IS NOT
// -----------------------------
//   * a refusal is shown verbatim, never rephrased and never summarised;
//   * `why` is shown beside every terminal state, because a reader of one
//     job does not have the workdir and cannot re-run the decision;
//   * a cancelled job's outputs are marked partial, from the record rather
//     than inferred from the state;
//   * no supervisor running is a fact the page states, not a button that
//     fails when pressed.
(() => {
  'use strict';
  const TOKEN = document.querySelector(
    'meta[name="oneground-lab-token"]').content;

  let sup = null;          // where the supervisor is, or why it is not
  let spec = null;         // stages, and what is never a job
  let seq = 0;             // newest refresh wins
  let timer = null;

  function el(tag, cls, text) {
    const n = document.createElement(tag);
    if (cls) n.className = cls;
    if (text !== undefined && text !== null) n.textContent = text;
    return n;
  }

  async function get(path, params) {
    const q = new URLSearchParams(params || {});
    q.set('token', TOKEN);
    const r = await fetch(path + '?' + q.toString(),
                          { headers: { 'Accept': 'application/json' } });
    const body = await r.json().catch(() => ({ error: 'unreadable answer' }));
    if (!r.ok) throw new Error(body.error || (path + ' answered ' + r.status));
    return body;
  }

  // The supervisor, directly. Not through the lab server, which makes no
  // request of its own.
  async function ask(path, body) {
    if (!sup || !sup.running) throw new Error(sup ? sup.why : 'no supervisor');
    const r = await fetch(sup.url + path, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Oneground-Supervisor-Token': sup.token,
      },
      body: JSON.stringify(body),
    });
    const out = await r.json().catch(() => ({
      error: 'the supervisor answered ' + r.status + ' and no readable body',
    }));
    if (!r.ok) { out.__status = r.status; throw out; }
    return out;
  }

  // --- drawing ------------------------------------------------------------
  function stateChip(job) {
    const chip = el('span', 'state state-' + job.state, job.state);
    if (job.partial) chip.appendChild(el('span', 'partial', ' partial'));
    return chip;
  }

  function jobRow(job) {
    const row = el('section', 'job');
    row.dataset.job = job.id;

    const head = el('div', 'job-head');
    head.appendChild(el('span', 'job-stage', job.stage));
    head.appendChild(stateChip(job));
    head.appendChild(el('span', 'job-when',
      job.ended || job.started || 'not started'));
    row.appendChild(head);

    // The invocation, which is what everything else hangs off: it is what
    // the receipts name and what a replay re-runs.
    const inv = el('pre', 'job-invocation',
      'oneground ' + (job.invocation || []).join(' '));
    row.appendChild(inv);

    // Why it is in the state it is in. Shown for every terminal job, not
    // only the interesting ones: a state without its reason hands over the
    // conclusion and keeps the evidence.
    if (job.why) row.appendChild(el('p', 'job-why', job.why));

    if (job.refusal) {
      // Verbatim. The most useful thing the tool produces, written once, in
      // the CLI, on purpose.
      const box = el('div', 'refusal');
      box.appendChild(el('p', 'refusal-what', job.refusal));
      row.appendChild(box);
    }

    const foot = el('div', 'job-foot');
    if (job.exit_code !== null && job.exit_code !== undefined) {
      foot.appendChild(el('span', 'job-exit', 'exit ' + job.exit_code));
    }
    if (job.build) foot.appendChild(el('span', 'job-build', job.build));

    const log = el('button', 'linkish', 'log');
    log.type = 'button';
    log.addEventListener('click', () => { showLog(job, row); });
    foot.appendChild(log);

    const live = job.state === 'queued' || job.state === 'running';
    if (live && sup && sup.running) {
      const stop = el('button', 'linkish', 'cancel');
      stop.type = 'button';
      stop.addEventListener('click', () => { cancel(job); });
      foot.appendChild(stop);
    }
    row.appendChild(foot);
    return row;
  }

  async function showLog(job, row) {
    const existing = row.querySelector('.job-log');
    if (existing) { existing.remove(); return; }
    const pre = el('pre', 'job-log', 'reading…');
    row.appendChild(pre);
    try {
      const out = await get('/api/jobs/log', { id: job.id });
      pre.textContent = out.text === null ? out.absent : out.text;
      if (out.text === null) pre.classList.add('absent');
    } catch (e) {
      pre.textContent = e.message;
    }
  }

  async function cancel(job) {
    try {
      await ask('/cancel', { id: job.id });
    } catch (e) {
      fail(e.refusal || e.error || String(e));
    }
    await refresh();
  }

  // THE CLICK ASKS. It does not default, and there is no pre-selected
  // target anywhere on this page.
  //
  // It used to send `[stage, '.']`, so clicking `characterize` enqueued
  // `oneground characterize .` against the runs directory -- measured: exit
  // 1, `PermissionError: [Errno 13] Permission denied: '.'`, a traceback,
  // for a job whose target the user was never asked for.
  //
  // **A job records exactly what ran, so a target nobody chose makes the
  // record accurate and meaningless** -- the whole value of the record gone
  // at the first click. A list with a default would have the same defect
  // wearing a control: the default is the thing that just bit us.
  //
  // So the target is part of the request, which is what the record has to
  // carry anyway, and **a stage whose target cannot be named is not
  // offered** -- with the reason, because a missing button is otherwise
  // indistinguishable from an oversight.
  let targets = null;

  function chooser(stage, spec_, anchor) {
    const existing = document.getElementById('chooser');
    if (existing) existing.remove();

    const box = el('section', 'chooser');
    box.id = 'chooser';
    box.appendChild(el('h4', null, 'Run ' + stage + ' against'));
    box.appendChild(el('p', 'note',
      (targets.kinds || {})[spec_.kind] || spec_.kind));

    const list = el('div', 'chooser-list');
    spec_.choices.forEach((choice) => {
      const b = el('button', 'door', choice.name);
      b.type = 'button';
      b.addEventListener('click', () => {
        box.remove();
        start(stage, choice.arg);
      });
      list.appendChild(b);
    });
    box.appendChild(list);

    // The command this would run, before it runs. Nothing is enqueued from
    // this page that the page has not already shown whole.
    const shown = el('pre', 'chooser-preview',
      'oneground ' + stage + ' <' + spec_.kind + '>');
    box.appendChild(shown);
    Array.prototype.forEach.call(list.querySelectorAll('.door'), (b, i) => {
      b.addEventListener('mouseenter', () => {
        shown.textContent = 'oneground ' + stage + ' '
          + spec_.choices[i].arg;
      });
    });

    const cancel_ = el('button', 'linkish', 'not now');
    cancel_.type = 'button';
    cancel_.addEventListener('click', () => { box.remove(); });
    box.appendChild(cancel_);

    anchor.appendChild(box);
  }

  async function start(stage, arg) {
    clearFail();
    try {
      // The invocation is exactly what a terminal would run, which is what
      // makes the record replayable.
      await ask('/enqueue', { stage: stage, invocation: [stage, arg] });
    } catch (e) {
      fail(e.refusal || e.error || String(e));
    }
    await refresh();
  }

  function fail(message) {
    const box = document.getElementById('jobs-error');
    if (!box) return;
    box.hidden = false;
    box.textContent = '';
    box.appendChild(el('p', 'refusal-what', message));
  }

  function clearFail() {
    const box = document.getElementById('jobs-error');
    if (box) { box.hidden = true; box.textContent = ''; }
  }

  // The supervisor's absence, stated. Not a disabled button and not a
  // button that fails: a sentence saying what is not running and what that
  // means for this page.
  function supervisorNote() {
    const note = el('section', 'supervisor-note');
    if (sup && sup.running && sup.same_build) {
      note.appendChild(el('p', 'note',
        'A supervisor is running and this page speaks to it directly. '
        + 'The server that drew this page starts nothing and writes '
        + 'nothing.'));
      return note;
    }
    note.classList.add('absent');
    note.appendChild(el('h3', null,
      sup && sup.running ? 'The supervisor is running a different build'
                         : 'No supervisor is running'));
    note.appendChild(el('p', 'note', (sup && sup.why)
      || 'nothing can be started from here'));
    if (!(sup && sup.running)) {
      note.appendChild(el('p', 'note',
        'Jobs already recorded are listed below and are read from the '
        + 'directory, so what has run is still visible.'));
    }
    return note;
  }

  function runner() {
    const box = el('section', 'runner');
    box.appendChild(el('h3', null, 'Start a stage'));

    const offered = (targets && targets.offered) || {};
    const names = Object.keys(offered);
    if (!names.length) {
      box.appendChild(el('p', 'note cannot-run',
        'No stage can be offered here yet: nothing in this directory can be '
        + 'named as a target for one.'));
    }

    // Two groups, because `pod plan` and `characterize` are different kinds
    // of thing and one undifferentiated row invites reading them as
    // equivalent.
    const groups = [
      ['The pipeline', names.filter((s) => s.indexOf('pod ') !== 0)],
      ['A pod session', names.filter((s) => s.indexOf('pod ') === 0)],
    ];
    groups.forEach(([title, stages]) => {
      if (!stages.length) return;
      box.appendChild(el('h4', 'runner-group', title));
      const row = el('div', 'runner-row');
      stages.forEach((stage) => {
        const b = el('button', 'door', stage + '…');
        b.type = 'button';
        b.disabled = !(sup && sup.running);
        // The ellipsis is the promise: this asks before it does anything.
        b.addEventListener('click', () => {
          chooser(stage, offered[stage], box);
        });
        row.appendChild(b);
      });
      box.appendChild(row);
    });

    // What is not offered, and why. A missing button is otherwise
    // indistinguishable from an oversight.
    ((targets && targets.withheld) || []).forEach((w) => {
      const p = el('p', 'note');
      p.appendChild(el('code', null, 'oneground ' + w.stage));
      p.appendChild(document.createTextNode(' is not offered: ' + w.why));
      box.appendChild(p);
    });

    // And what is never a job at all, which is a ruling rather than a gap.
    (spec.not_a_job || []).forEach((n) => {
      const p = el('p', 'note');
      p.appendChild(el('code', null, 'oneground ' + n.stage));
      p.appendChild(document.createTextNode(
        ' is never a job here: ' + n.why + '. The pod card shows the '
        + 'command to run at a terminal.'));
      box.appendChild(p);
    });
    return box;
  }

  async function refresh() {
    const mine = ++seq;
    clearFail();
    let out;
    try {
      out = await get('/api/jobs');
    } catch (e) {
      if (mine !== seq) return;
      fail(e.message);
      return;
    }
    if (mine !== seq) return;
    spec = out;
    try {
      targets = await get('/api/jobs/targets');
    } catch (e) {
      targets = { offered: {}, withheld: [], kinds: {} };
      fail(e.message);
    }
    if (mine !== seq) return;
    const list = document.getElementById('jobs-list');
    if (!list) return;
    list.textContent = '';
    if (!out.jobs.length) {
      list.appendChild(el('p', 'note',
        'Nothing has been run from this directory yet.'));
    }
    out.jobs.slice().reverse().forEach((job) => {
      list.appendChild(jobRow(job));
    });
    const host = document.getElementById('jobs-runner');
    if (host) { host.textContent = ''; host.appendChild(runner()); }
  }

  async function showJobs() {
    const main = document.getElementById('ui-main');
    main.textContent = '';
    sup = await get('/api/supervisor');
    spec = await get('/api/jobs');

    main.appendChild(el('h2', null, 'Jobs'));
    main.appendChild(el('p', 'lede',
      'Every stage is a command. What ran, what it exited with, why it is '
      + 'in the state it is in, and the log it wrote — all of it read from '
      + 'this directory, which is the truth.'));
    main.appendChild(supervisorNote());

    const runnerHost = el('div');
    runnerHost.id = 'jobs-runner';
    main.appendChild(runnerHost);

    const err = el('div', 'refusal');
    err.id = 'jobs-error';
    err.hidden = true;
    err.setAttribute('role', 'alert');
    main.appendChild(err);

    const list = el('div');
    list.id = 'jobs-list';
    main.appendChild(list);

    await refresh();

    // A running job's state changes without anyone clicking. Cleared when
    // the page is left, so a route away does not leave a timer redrawing a
    // page that is gone.
    if (timer) window.clearInterval(timer);
    timer = window.setInterval(() => {
      if (window.location.hash !== '#/jobs') {
        window.clearInterval(timer);
        timer = null;
        return;
      }
      refresh();
    }, 4000);
  }

  window.onegroundJobs = showJobs;
})();
